### PR TITLE
change imsave to use libtiff

### PR DESCRIPTION
**Pull request by sunyi000**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [x] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

we think (and also tested with Fiji) the original matplotlib imsave function saves all channels instead of just one, even though just one channel id is passed into the getTile function.

with the new library libtiff, it just saves the image from the channel specified.

thanks very much for the libtiff suggestion @beatrizserrano 

Q: Do we also need to modify the python-omero recipe to include libtiff ?

### CODE REVIEW

File: tools/idr_download/idr_download_by_ids.py [hunk 1]
```diff
@@ -86,7 +87,9 @@ def download_plane_as_tiff(image, tile, z, c, t, fname):
 
     if fname[-5:] != '.tiff':
         fname += '.tiff'
-    plt.imsave(fname, selection)
+    tiff = TIFF.open(fname, mode = 'w')
```
Comment (bgruening) [inline tools/idr_download/idr_download_by_ids.py]: does this support `with TIFF.open(fname, mode = 'w') as tiff:` ?

Comment (sunyi000) [reply tools/idr_download/idr_download_by_ids.py]: i had a peek at the source code, it doesnt seem to support this.. unless we want to hack into its 'open' function and copy some stuff into here

File: tools/idr_download/idr_download_by_ids.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import argparse
 import os
 import sys
+from libtiff import TIFF
 
 from matplotlib import pyplot as plt
```
Comment (beatrizserrano) [inline tools/idr_download/idr_download_by_ids.py]:
```suggestion

```

**Comment by sunyi000** [comment]
@bgruening I see the linting error in the travisCi ouput below, however my local planemo returns all green.. I thought the lint should be green even without updating the recipe?

```
./tools/idr_download/idr_download_by_ids.py:90:33: E251 unexpected spaces around keyword / parameter equals
```

File: tools/idr_download/idr_download_by_ids.py (hunk 2 above)
Comment (sunyi000) [reply tools/idr_download/idr_download_by_ids.py]: ah ye. this should be removed thx

**Comment by bgruening** [comment]
@sunyi000 this is a pep8 warning. So you python code is not following pep8. This is different from planemo. Its really a python-only thing. You can run locally flake8 on your python file if you like.

**Comment by sunyi000** [comment]
the libtiff in conda-forge seems to be different. 
I used this one here which is a wrapper for the c library.
https://pypi.org/project/libtiff/

it has the same name 'libtiff' so it creates some problems, so I just ''pip install'' within the omero env. 
I know this is not the right way, but since it has name conflits..I'm not sure how to fix yet..
any help are appreciated @bgruening

**Comment by bgruening** [comment]
@sunyi000 see here: https://github.com/conda-forge/staged-recipes/pull/11145

**Comment by jburel** [comment]
@beatrizserrano @sunyi000 are you using that branch to connect to IDR or still the previous code?
we are noticing still loads of File 
``
"/opt/omero/web/venv3/lib64/python3.6/site-packages/omero/gateway/__init__.py", line 8056, in _prepareRenderingEngine 
`` 
this should be fixed by  commit e39af0e

**Comment by sunyi000** [comment]
> @beatrizserrano @sunyi000 are you using that branch to connect to IDR or still the previous code?
> we are noticing still loads of File
> `"/opt/omero/web/venv3/lib64/python3.6/site-packages/omero/gateway/__init__.py", line 8056, in _prepareRenderingEngine `
> this should be fixed by commit [e39af0e](https://github.com/galaxyproject/tools-iuc/commit/e39af0e77e0f6a3aa0965d2bb47cc170ef96c693)

Hi @jburel ,

Is IDR running today.. I can't seem to download any files today.. it seems not even connecting to the database.

- I did try downloading lots of files since last Friday and I got various errors for some of the images.  But most images I can download successfully

- today seems nothing is downloaded..

thanks

**Comment by jburel** [comment]
It is up
unless it is coming from something else

**Comment by sunyi000** [comment]
thanks @jburel 
here is what I have experienced since yesterday afternoon. appreciate any help or pointers 

- I took this idr_download_by_id.py and run it in python-omero conda env. it was running with some errors before but I still can download the majority of images. 

- since yesterday afternoon, I put a "print" after the blizgateway connection, just to debug the connection, the code doesn't seem to connect

- i try using the tool from galaxyeu, just to download 1 single image, and the job is running forever..

thanks for looking into it..

**Comment by beatrizserrano** [comment]
I've also checked in the website and many of the thumbnails are displayed as "No image". Also when I try to download one that is shown correctly I see:
```
Server Error. (500)
The server encountered an internal error and was unable to complete your request. Please give us feedback to help to improve the OMERO applications. Thanks for your patience.
```

More specifically:
```
File "/opt/omero/web/venv3/lib64/python3.6/site-packages/omero/gateway/__init__.py", line 7785, in wrapped
    except ctx.ignoreExceptions:

TypeError: catching classes that do not inherit from BaseException is not allowed

<WSGIRequest: GET '/webclient/render_image_download/223928/?format=tif
```

**Comment by jburel** [comment]
The temporary infrastructure we are running IDR on during the Embassy cloud upgrade is having few issues. 
The upgrade is a bit delay so we will unfortunately be back to normal at the end of this month as expected. 
The current global pandemic does not help to get things go back to normal quickly either
I hope you will understand.

**Comment by sunyi000** [comment]
if just run the 5145th image, id=2170646, the exception is below.
```
echo 2170646 | python idr_download_by_ids.py -c 'Hoechst' -f 0 -z 0 --rectangle 0 0 696 520 --skip-failed
```

if run all the id files at once.. some session closed exception occured..i will need to re-run it to find the exact exception for this one. 

```
WARNING:omero.gateway:ResourceError on <class 'omero.gateway.OmeroGatewaySafeCallWrapper'> to <7f3498a4-a5b2-469b-9961-c8354d723d69omero.api.RawPixelsStore> setPixelsId((2170646, True, <ServiceOptsDict: {'omero.client.uuid': '56be61e8-e8ec-4f50-bb14-3538d62977f6', 'omero.event': 'Internal', 'omero.session.uuid': '5912537b-7bbd-4ddd-8e9b-e7949d08250a'}>), {})
Traceback (most recent call last):
  File "/home/ysun/miniconda2/envs/idr/lib/python3.8/site-packages/omero/gateway/__init__.py", line 4746, in __call__
    return self.f(*args, **kwargs)
  File "/home/ysun/miniconda2/envs/idr/lib/python3.8/site-packages/omero_api_RawPixelsStore_ice.py", line 529, in setPixelsId
    return _M_omero.api.RawPixelsStore._op_setPixelsId.invoke(self, ((pixelsId, bypassOriginalFile), _ctx))
omero.ResourceError: exception ::omero::ResourceError
{
    serverStackTrace = ome.conditions.ResourceError: Error instantiating pixel buffer: /data/OMERO/ManagedRepository/demo_2/2016-06/17/19-59-29.574/metadata/idr0016-wawer-bioactivecompoundprofiling/screens/24753.screen
        at ome.io.nio.PixelsService.createBfPixelBuffer(PixelsService.java:877)
        at ome.io.nio.PixelsService._getPixelBuffer(PixelsService.java:631)
        at ome.io.nio.PixelsService.getPixelBuffer(PixelsService.java:549)
        at ome.services.RawPixelsBean.setPixelsId(RawPixelsBean.java:284)
        at sun.reflect.GeneratedMethodAccessor908.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:333)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
        at ome.security.basic.EventHandler.invoke(EventHandler.java:154)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at ome.tools.hibernate.SessionHandler.doStateful(SessionHandler.java:216)
        at ome.tools.hibernate.SessionHandler.invoke(SessionHandler.java:200)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
        at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:282)
        at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:249)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:213)
        at com.sun.proxy.$Proxy101.setPixelsId(Unknown Source)
        at sun.reflect.GeneratedMethodAccessor908.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:333)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
        at ome.security.basic.BasicSecurityWiring.invoke(BasicSecurityWiring.java:93)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at ome.services.blitz.fire.AopContextInitializer.invoke(AopContextInitializer.java:43)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:213)
        at com.sun.proxy.$Proxy101.setPixelsId(Unknown Source)
        at sun.reflect.GeneratedMethodAccessor991.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at ome.services.blitz.util.IceMethodInvoker.invoke(IceMethodInvoker.java:172)
        at ome.services.throttling.Callback.run(Callback.java:56)
        at ome.services.throttling.InThreadThrottlingStrategy.callInvokerOnRawArgs(InThreadThrottlingStrategy.java:56)
        at ome.services.blitz.impl.AbstractAmdServant.callInvokerOnRawArgs(AbstractAmdServant.java:140)
        at ome.services.blitz.impl.RawPixelsStoreI.setPixelsId_async(RawPixelsStoreI.java:197)
        at sun.reflect.GeneratedMethodAccessor990.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:333)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
        at omero.cmd.CallContext.invoke(CallContext.java:85)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:213)
        at com.sun.proxy.$Proxy102.setPixelsId_async(Unknown Source)
        at omero.api._RawPixelsStoreTie.setPixelsId_async(_RawPixelsStoreTie.java:270)
        at omero.api._RawPixelsStoreDisp.___setPixelsId(_RawPixelsStoreDisp.java:668)
        at omero.api._RawPixelsStoreDisp.__dispatch(_RawPixelsStoreDisp.java:1759)
        at IceInternal.Incoming.invoke(Incoming.java:221)
        at Ice.ConnectionI.invokeAll(ConnectionI.java:2536)
        at Ice.ConnectionI.dispatch(ConnectionI.java:1145)
        at Ice.ConnectionI.message(ConnectionI.java:1056)
        at IceInternal.ThreadPool.run(ThreadPool.java:395)
        at IceInternal.ThreadPool.access$300(ThreadPool.java:12)
        at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:832)
        at java.lang.Thread.run(Thread.java:748)

    serverExceptionClass = ome.conditions.ResourceError
    message = Error instantiating pixel buffer: /data/OMERO/ManagedRepository/demo_2/2016-06/17/19-59-29.574/metadata/idr0016-wawer-bioactivecompoundprofiling/screens/24753.screen
}
ERROR:omero.gateway:Failed to getPlane() or getTile() from rawPixelsStore
Traceback (most recent call last):
  File "/home/ysun/miniconda2/envs/idr/lib/python3.8/site-packages/omero/gateway/__init__.py", line 7361, in getTiles
    rawPixelsStore = self._prepareRawPixelsStore()
  File "/home/ysun/miniconda2/envs/idr/lib/python3.8/site-packages/omero/gateway/__init__.py", line 7258, in _prepareRawPixelsStore
    ps.setPixelsId(self._obj.id.val, True, self._conn.SERVICE_OPTS)
  File "/home/ysun/miniconda2/envs/idr/lib/python3.8/site-packages/omero/gateway/__init__.py", line 4749, in __call__
    return self.handle_exception(e, *args, **kwargs)
  File "/home/ysun/miniconda2/envs/idr/lib/python3.8/site-packages/omero/gateway/__init__.py", line 4746, in __call__
    return self.f(*args, **kwargs)
  File "/home/ysun/miniconda2/envs/idr/lib/python3.8/site-packages/omero_api_RawPixelsStore_ice.py", line 529, in setPixelsId
    return _M_omero.api.RawPixelsStore._op_setPixelsId.invoke(self, ((pixelsId, bypassOriginalFile), _ctx))
omero.ResourceError: exception ::omero::ResourceError
{
    serverStackTrace = ome.conditions.ResourceError: Error instantiating pixel buffer: /data/OMERO/ManagedRepository/demo_2/2016-06/17/19-59-29.574/metadata/idr0016-wawer-bioactivecompoundprofiling/screens/24753.screen
        at ome.io.nio.PixelsService.createBfPixelBuffer(PixelsService.java:877)
        at ome.io.nio.PixelsService._getPixelBuffer(PixelsService.java:631)
        at ome.io.nio.PixelsService.getPixelBuffer(PixelsService.java:549)
        at ome.services.RawPixelsBean.setPixelsId(RawPixelsBean.java:284)
        at sun.reflect.GeneratedMethodAccessor908.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:333)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
        at ome.security.basic.EventHandler.invoke(EventHandler.java:154)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at ome.tools.hibernate.SessionHandler.doStateful(SessionHandler.java:216)
        at ome.tools.hibernate.SessionHandler.invoke(SessionHandler.java:200)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
        at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:282)
        at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:249)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:213)
        at com.sun.proxy.$Proxy101.setPixelsId(Unknown Source)
        at sun.reflect.GeneratedMethodAccessor908.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:333)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
        at ome.security.basic.BasicSecurityWiring.invoke(BasicSecurityWiring.java:93)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at ome.services.blitz.fire.AopContextInitializer.invoke(AopContextInitializer.java:43)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:213)
        at com.sun.proxy.$Proxy101.setPixelsId(Unknown Source)
        at sun.reflect.GeneratedMethodAccessor991.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at ome.services.blitz.util.IceMethodInvoker.invoke(IceMethodInvoker.java:172)
        at ome.services.throttling.Callback.run(Callback.java:56)
        at ome.services.throttling.InThreadThrottlingStrategy.callInvokerOnRawArgs(InThreadThrottlingStrategy.java:56)
        at ome.services.blitz.impl.AbstractAmdServant.callInvokerOnRawArgs(AbstractAmdServant.java:140)
        at ome.services.blitz.impl.RawPixelsStoreI.setPixelsId_async(RawPixelsStoreI.java:197)
        at sun.reflect.GeneratedMethodAccessor990.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:333)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
        at omero.cmd.CallContext.invoke(CallContext.java:85)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
        at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:213)
        at com.sun.proxy.$Proxy102.setPixelsId_async(Unknown Source)
        at omero.api._RawPixelsStoreTie.setPixelsId_async(_RawPixelsStoreTie.java:270)
        at omero.api._RawPixelsStoreDisp.___setPixelsId(_RawPixelsStoreDisp.java:668)
        at omero.api._RawPixelsStoreDisp.__dispatch(_RawPixelsStoreDisp.java:1759)
        at IceInternal.Incoming.invoke(Incoming.java:221)
        at Ice.ConnectionI.invokeAll(ConnectionI.java:2536)
        at Ice.ConnectionI.dispatch(ConnectionI.java:1145)
        at Ice.ConnectionI.message(ConnectionI.java:1056)
        at IceInternal.ThreadPool.run(ThreadPool.java:395)
        at IceInternal.ThreadPool.access$300(ThreadPool.java:12)
        at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:832)
        at java.lang.Thread.run(Thread.java:748)

    serverExceptionClass = ome.conditions.ResourceError
    message = Error instantiating pixel buffer: /data/OMERO/ManagedRepository/demo_2/2016-06/17/19-59-29.574/metadata/idr0016-wawer-bioactivecompoundprofiling/screens/24753.screen
}
ImageSpecWarning for 24753 [Well N10, Field 1] (ID: 2170646): Could not download the requested region
```

**Comment by beatrizserrano** [comment]
>     message = Error instantiating pixel buffer: /data/OMERO/ManagedRepository/demo_2/2016-06/17/19-59-29.574/metadata/idr0016-wawer-bioactivecompoundprofiling/screens/24753.screen

This is the error when an image is missing, see here: http://idr.openmicroscopy.org/webclient/img_detail/2170646/

And the issue reported here: https://github.com/idr-contrib/community/issues/5

**Comment by sunyi000** [comment]
sorry I seemed to reply in the wrong place...

**Comment by beatrizserrano** [comment]
@jburel @sunyi000 do you see anything else here to change? If not, maybe @bgruening can merge :)

**Comment by bgruening** [comment]
Thanks all!

**Pull request merged**